### PR TITLE
Fix the memory leaks

### PR DESCRIPTION
**Pull request by rozap**
* Add backpressure to kmz stream
* Add backpressure to kml stream
* Add backpressure to shapefile stream
* Abandoned the JSONStream library
  * The underlying JSON parser held onto the root object
    rather than just emitting events, leading to memory
    issues
  * Added clarinet library for sax style json parsing
  * Added util/parser.js stream transform which exposes
    a similar api to JSONStream but shouldn't eat up memory
  * Simple tests for util/parser.js
  * parser.js should be pulled out into an npm module
* Added flow control tests
  * Stream large-ish fixtures through the decoders and check
    to make sure the internal buffer never exceeds a threshold
* Added Metrics class for heapdumps when lots of memory is
  allocated

### CODE REVIEW

File: lib/config/config.js [hunk 1]
```diff
@@ -11,7 +11,7 @@ export default {
   //so if we have 100 layers (this never actually happens irl),
```
Comment (michaelb990) [inline lib/config/config.js]: Update this comment. 20 layers seems like it may happen?

File: lib/decoders/shapefile.js [hunk 2]
```diff
@@ -1,5 +1,11 @@
 /**
  * Convert a shapefile into SoQL Values
+ *
+ * NOTE: this is not pausable! pause will
+ * put the stream into a paused state forever,
+ * and resume will not work.
+ * TODO: if this functionality is needed at some point,
+ * need to revisit it.
  */
 import JSONStream from 'JSONStream';
```
Comment (michaelb990) [inline lib/decoders/shapefile.js]: Think you need to remove this import.

File: lib/decoders/kmz.js [hunk 3]
```diff
@@ -49,6 +48,32 @@ class KMZ extends Duplex {
...
+    this.emit('readable');
+  }
+
+  _onOpenKmlStream(kmlStream, zipFile) {
+    kmlStream
+      .pipe(this._kmlDecoder)
+      .on('error', (err) => {
+        this.emit('error', err);
+      })
+      .on('data', (data) => {
```
Comment (michaelb990) [inline lib/decoders/kmz.js]: What is happening here? Why do we want a no-op if `this.push(data) == true`?

File: lib/decoders/merger.js [hunk 4]
```diff
@@ -67,13 +67,12 @@ class Merger extends Transform {
...
     var [newLayers, layer] = getOrCreateLayer(this._layers, chunk, this._disk, spec);
-    layer.write(chunk.crs, chunk.columns, this._throwaway);
     this._layers = newLayers;
 
-    return done();
+    layer.write(chunk.crs, chunk.columns, this._throwaway, done);
   }
 
-  end() {
+  end(cb) {
```
Comment (michaelb990) [inline lib/decoders/merger.js]: How is `cb` used?

**Comment by michaelb990** [comment]
A few questions, but it looks good (to me). 

I'd love to hear more about how the metrics class works. Maybe I could see an example of the output? Seems like a really useful tool for general use.

File: lib/decoders/layer.js [hunk 5]
```diff
@@ -192,25 +192,31 @@ class Layer extends Duplex {
...
    */
-  write(crs, soqlRow, throwaway) {
+  write(crs, soqlRow, throwaway, done) {
     if (crs) this._crsMap[this._count] = crs;
 
     this._updateColumnTypes(soqlRow);
     this._count++;
-    if (throwaway) return;
+    if (throwaway) {
+      if(done) done();
```
Comment (cyranix) [inline lib/decoders/layer.js]: \<nit> space after `if`

File: lib/services/spatial.js [hunk 6]
```diff
@@ -56,7 +56,7 @@ class SpatialService {
   _createLayers(req, res, core, layers) {
     async.mapLimit(layers, MAX_PARALLEL, core.create.bind(core), (err, datasetResponses) => {
       //failed to get upstream from zk or create request failed
-      if (err) return res.status(err.statusCode).send(err.body);
+      if (err) return res.status(err.statusCode || 500).send(err.body);
```
Comment (cyranix) [inline lib/services/spatial.js]: Is `err.body` guaranteed to exist even if `err.statusCode` isn't?

File: lib/decoders/kmz.js (hunk 3 above)
Comment (cyranix) [inline lib/decoders/kmz.js]: Seems like a backpressure mechanism? If we fail to do the expected thing (push the data), wait for things to clear up before continuing to process.

Comment (rozap) [inline lib/decoders/kmz.js]: `push` has a side effect of pushing the datum onto the stream, so it's not a noop. It's just that if `push(data) == true` then that means the stream's high water mark hasn't been hit, so we can keep pushing stuff. If push is `false` then that means the stream is being overwhelmed and we need to pause it and resume it when a `drain` event is fired.

File: lib/decoders/merger.js (hunk 4 above)
Comment (rozap) [inline lib/decoders/merger.js]: looks like it's not, i'll remove it

File: lib/services/spatial.js (hunk 6 above)
Comment (rozap) [inline lib/services/spatial.js]: nope, good catch

**Comment by rozap** [comment]
The metrics class doesn't really do anything right now, in that it's not hooked up to any socrata metrics services. 

It just takes the node process memory stats and logs them. A log line might look like this
`{"name":"geo-import","hostname":"chris-macbook","pid":21432,"level":30,"rss":158515200,"heapTotal":137338112,"heapUsed":100345032,"rssGrowth":1,"msg":"metrics","time":"2015-12-28T18:57:09.819Z","v":0}`

It will also make a heap dump when the resident set size grows by a large amount between samples. It's really stupid, and should probably be cleaned up asap, as taking a heapdump will stop the world, but it is useful in the early stages of this project for debugging purposes.

File: lib/decoders/kmz.js (hunk 3 above)
Comment (michaelb990) [inline lib/decoders/kmz.js]: :+1:

**Comment by michaelb990** [comment]
:shipit:

**Pull request merged**